diagnostics.go

```diff
@@ -0,0 +1,323 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pilosa
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+// Default version check URL.
+const (
+	defaultVersionCheckURL = "https://diagnostics.pilosa.com/v0/version"
+)
+
+type versionResponse struct {
+	Version string `json:"version"`
+	Message string `json:"message"`
+}
+
+// DiagnosticsCollector represents a collector/sender of diagnostics data
+type DiagnosticsCollector struct {
+	mu          sync.Mutex
+	host        string
+	VersionURL  string
+	version     string
+	lastVersion string
+	startTime   int64
+	start       time.Time
+
+	metrics map[string]interface{}
+
+	client *http.Client
+
+	logOutput io.Writer
+
+	server *Server
+}
+
+// New returns a pointer to a new DiagnosticsCollector Client given an addr in the format "hostname:port".
+func NewDiagnosticsCollector(host string) *DiagnosticsCollector {
+
+	return &DiagnosticsCollector{
+		host:       host,
+		VersionURL: defaultVersionCheckURL,
+		startTime:  time.Now().Unix(),
+		start:      time.Now(),
+		client:     &http.Client{Timeout: 10 * time.Second},
+		metrics:    make(map[string]interface{}),
+		logOutput:  ioutil.Discard,
+	}
+}
+
+// SetVersion of locally running Pilosa Cluster to check against master.
+func (d *DiagnosticsCollector) SetVersion(v string) {
+	d.version = v
+	d.Set("Version", v)
+}
+
+// Flush sends the current metrics.
+func (d *DiagnosticsCollector) Flush() error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.metrics["Uptime"] = (time.Now().Unix() - d.startTime)
+	buf, err := d.encode()
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest("POST", d.host, bytes.NewReader(buf))
+	req.Header.Set("Content-Type", "application/json")
+	resp, err := d.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// TODO verify response
+	_, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// CheckVersion of the local build against Pilosa master.
+func (d *DiagnosticsCollector) CheckVersion() error {
+	var rsp versionResponse
+	req, err := http.NewRequest("GET", d.VersionURL, nil)
+	resp, err := d.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("http: status=%d", resp.StatusCode)
+	} else if err := json.NewDecoder(resp.Body).Decode(&rsp); err != nil {
+		return fmt.Errorf("json decode: %s", err)
+	}
+
+	// Same a version as last test
```

Comment:
the wording of this comment is confusing.

Reply:
Agreed, fixed.